File: Java/Binary Tree Paths.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> res = new ArrayList<String>();
        find(root, res, new StringBuilder());
        return res;
    }
    
    private void find(TreeNode root, List<String> res, StringBuilder sb) {
        if (root == null)
            return;
        int len = sb.length();
        if (root.left == null && root.right == null) {
            sb.append(root.val);
            res.add(sb.toString());
            sb.delete(len, sb.length());
            return;
        }
        sb.append(root.val);
        sb.append("->");
        
        find(root.left, res, sb);
        find(root.right, res, sb);
        sb.delete(len, sb.length());
    }
}

File: Python/Binary Tree Paths.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    """
    :type root: TreeNode
    :rtype: List[str]
    """
    def binaryTreePaths(self, root):
        res = []
        self.printPaths(root, res, [])
        return res
        
    def printPaths(self, root, res, path):
        if root == None:
            return
        if root.left == None and root.right == None:
            path.append(str(root.val))
            res.append("".join(path))
            del path[len(path)-1]
            return
        
        path.append(str(root.val))
        path.append("->")
        
        self.printPaths(root.left, res, path)
        self.printPaths(root.right, res, path)
        
        del path[len(path)-2:len(path)]
        